File: swarmmanagercontroller/src/main/java/co/uk/swarmbit/docker/api/swarm/SwarmApiImpl.java
package co.uk.swarmbit.docker.api.swarm;

import co.uk.swarmbit.docker.api.common.AbstractApiImpl;
import co.uk.swarmbit.docker.api.swarm.parameters.SwarmUpdateParameters;
import co.uk.swarmbit.docker.api.common.json.UnlockKeyJson;
import co.uk.swarmbit.docker.api.common.json.SwarmJson;
import co.uk.swarmbit.rest.RestExecutorFactory;
import co.uk.swarmbit.rest.RestMethod;
import co.uk.swarmbit.rest.RestParameters;
import co.uk.swarmbit.rest.RestResponseType;
import org.springframework.stereotype.Component;

@Component
public class SwarmApiImpl  extends AbstractApiImpl implements SwarmApi {

    private static final String SWARM_PATH = "/swarm";

    private static final String UNLOCK_PATH = "/unlock";

    private static final String UPDATE_PATH = "/update";

    @Override
    public SwarmJson inspectSwarm(String swarmId) {
        return inspectObject(SWARM_PATH, swarmId, new RestResponseType<SwarmJson>() {});
    }

    @Override
    public void updateSwarm(String swarmId, SwarmUpdateParameters parameters) {
        updateObject(SWARM_PATH + UPDATE_PATH, swarmId, new RestResponseType<Void>() {}, parameters, parameters);
    }

    @Override
    public UnlockKeyJson unlock(String swarmId) {
        RestParameters restParameters = new RestParameters(dockerWebClient.getBaseResource(swarmId))
                .setPath(SWARM_PATH + UNLOCK_PATH);
        return RestExecutorFactory.createRestExecutor(RestMethod.POST).execute(restParameters, new RestResponseType<UnlockKeyJson>() {});
    }
}

File: swarmmanagercontroller-new/src/main/kotlin/com/swarmbit/docker/api/swarm/SwarmApiImpl.kt
package com.swarmbit.docker.api.swarm

import com.swarmbit.docker.api.common.AbstractApiImpl
import com.swarmbit.docker.api.common.client.DockerWebClient
import com.swarmbit.docker.api.common.rest.executor.RestExecutorFactory
import com.swarmbit.docker.api.common.rest.model.Method
import com.swarmbit.docker.api.common.rest.model.Parameters
import com.swarmbit.docker.api.common.rest.model.ResponseType
import com.swarmbit.docker.api.swarm.model.Swarm
import com.swarmbit.docker.api.swarm.model.UnlockKey
import com.swarmbit.docker.api.swarm.parameters.SwarmUpdateParameters
import io.micronaut.context.annotation.Context

@Context
class SwarmApiImpl(private val dockerWebClient: DockerWebClient): AbstractApiImpl(dockerWebClient), SwarmApi {

    companion object {
        private const val SWARM_PATH = "/swarm"
        private const val UNLOCK_PATH = "/unlock"
        private const val UPDATE_PATH = "/update"
    }

    override fun inspectSwarm(swarmId: String): Swarm {
        return inspectObject(SWARM_PATH, swarmId, object : ResponseType<Swarm>() {})
                ?: throw IllegalArgumentException("Swarm not found for swarmId ($swarmId)")
    }

    override fun updateSwarm(swarmId: String, parameters: SwarmUpdateParameters) {
        updateObject(SWARM_PATH + UPDATE_PATH, swarmId, object : ResponseType<Void>() {}, parameters, parameters)
    }

    override fun unlock(swarmId: String): UnlockKey {
        val restParameters: Parameters = Parameters(dockerWebClient.getBaseResource(swarmId)).setPath(SWARM_PATH + UNLOCK_PATH)
        return RestExecutorFactory.createRestExecutor(Method.POST).execute(restParameters, object : ResponseType<UnlockKey>() {})!!
    }

}